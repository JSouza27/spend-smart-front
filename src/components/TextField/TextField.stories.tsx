import { ComponentMeta, ComponentStory } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import { zodResolver } from '@hookform/resolvers/zod';
import * as z from 'zod';

import TextField from '.';
import theme from '../../styles/theme';
import { FormProvider, useForm } from 'react-hook-form';

export default {
  title: 'TextField',
  component: TextField,
  args: {
    label: 'Description',
    name: 'description'
  }
} as ComponentMeta<typeof TextField>;

const schema = z.object({
  description: z.string().max(3, 'Erro nesse input')
});

type SchemaFormValues = z.infer<typeof schema>;

export const Basic: ComponentStory<typeof TextField> = (args) => {
  const methods = useForm();

  return (
    <ThemeProvider theme={theme}>
      <FormProvider {...methods}>
        <TextField {...args} />
      </FormProvider>
    </ThemeProvider>
  );
};

export const Disabled: ComponentStory<typeof TextField> = (args) => {
  const methods = useForm();

  return (
    <ThemeProvider theme={theme}>
      <FormProvider {...methods}>
        <TextField disabled value="Estou desabilitado" {...args} />
      </FormProvider>
    </ThemeProvider>
  );
};
