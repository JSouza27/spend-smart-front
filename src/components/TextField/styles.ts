import styled, { DefaultTheme, css } from 'styled-components';
import { TextFieldProps } from '.';

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    gap: ${theme.spacings.xxsmall};
    width: 100%;
  `}
`;

type LabelProps = Pick<TextFieldProps, 'required' | 'disabled'>;

export const Label = styled.label<LabelProps>`
  ${({ theme, required, disabled }) => css`
    color: ${disabled ? theme.colors.neutral_600 : theme.colors.neutral_700};
    font-size: ${theme.font.sizes.xsmall};
    line-height: ${theme.font.sizes.medium};

    &::before {
      content: ${required ? '*' : ''};
      color: ${theme.colors.red_400};
    }
  `}
`;

type InputProps = {
  isError: boolean;
};

const inputModifiers = {
  true: (theme: DefaultTheme) => css`
    border: 1px solid;
    border-color: ${theme.colors.red_400};
    color: ${theme.colors.neutral_800};

    &:focus {
      border-color: ${theme.colors.red_300};
      box-shadow: 0 0 0 2px ${theme.colors.red_200};
      outline-color: ${theme.colors.red_300};
    }
  `,
  false: (theme: DefaultTheme) => css`
    border: 1px solid;
    border-color: ${theme.colors.neutral_400};
    color: ${theme.colors.neutral_600};

    &:focus {
      border-color: ${theme.colors.blue_300};
      box-shadow: 0 0 0 2px ${theme.colors.blue_200};
      outline-color: ${theme.colors.blue_300};
    }
  `
};

export const Input = styled.input<InputProps>`
  ${({ theme, isError }) => css`
    background-color: ${theme.colors.neutral_0};
    border-radius: ${theme.border.radius};
    font-size: ${theme.font.sizes.xsmall};
    padding: 1.2rem ${theme.spacings.xxsmall};
    width: 100%;

    ${isError && inputModifiers.true(theme)};
    ${!isError && inputModifiers.false(theme)};

    &:disabled {
      background-color: ${theme.colors.neutral_100};
      border: 1px solid ${theme.colors.neutral_300};
      color: ${theme.colors.neutral_600};
    }
  `}
`;
