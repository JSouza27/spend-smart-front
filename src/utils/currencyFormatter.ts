export class CurrencyFormatter {
  private static language = 'pt-BR';
  private static currency = 'BRL';

  static formatter(num: number): string {
    return (num ?? 0).toLocaleString(this.language, {
      currency: this.currency
    });
  }

  static unformat(num: string): number {
    const nums = num.replace(/[^0-9]+/g, '');
    const i = nums.length - 2;

    return Number(`${nums.substring(0, i)}.${nums.substring(i)}`);
  }
}
