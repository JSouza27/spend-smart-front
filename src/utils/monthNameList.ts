export const monthNameList = [
  'janeiro',
  'fevereiro',
  'mar√ßo',
  'abril',
  'maio',
  'junho',
  'julho',
  'agosto',
  'setembro',
  'outubro',
  'novembro',
  'dezembro'
];
