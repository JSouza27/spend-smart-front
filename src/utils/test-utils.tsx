import { render } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../styles/theme';
import { FormProvider, useForm } from 'react-hook-form';

const AllTheProviders = ({ children }: any) => {
  const methods = useForm();

  return (
    <ThemeProvider theme={theme}>
      <FormProvider {...methods}>{children}</FormProvider>
    </ThemeProvider>
  );
};

const customRender = (
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  options?: any
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
