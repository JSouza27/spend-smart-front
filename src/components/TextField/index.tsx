import { InputHTMLAttributes } from 'react';
import { FieldValues, RegisterOptions, useFormContext } from 'react-hook-form';

import * as S from './styles';
import InvalidFeedBack from '../InvalidFeedBack';

export type TextFieldProps = {
  label: string;
  name: string;
  options?: RegisterOptions<FieldValues, string> | undefined;
} & InputHTMLAttributes<HTMLInputElement>;

export default function TextField({
  label,
  name,
  required,
  options,
  ...props
}: TextFieldProps) {
  const {
    register,
    formState: { errors }
  } = useFormContext();

  const isError = Object.keys(errors).includes(name);
  const message = errors[name]?.message;

  return (
    <S.Wrapper>
      <S.Label htmlFor={name}>{label}</S.Label>
      <S.Input
        id={name}
        isError={isError}
        type="text"
        {...register(name, options)}
        {...props}
      />
      {errors[name] && message !== undefined && (
        <InvalidFeedBack message={message} />
      )}
    </S.Wrapper>
  );
}
