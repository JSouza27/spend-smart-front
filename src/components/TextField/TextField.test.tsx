import { fireEvent, render, screen } from 'utils/test-utils';

import TextField from '.';

describe('<TextField />', () => {
  it('should render the TextField', () => {
    const { container } = render(
      <TextField label="Descrição" name="description" type="text" />
    );

    const label = screen.getByRole('textbox', {
      name: /descrição/i
    });
    const input = screen.getByRole('textbox', {
      name: /descrição/i
    });

    fireEvent.change(input, {
      target: { value: 'Test transaction' }
    });

    expect(label).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(input).toHaveProperty('value', 'Test transaction');
    expect(container.firstChild).toMatchSnapshot();
  });
});
