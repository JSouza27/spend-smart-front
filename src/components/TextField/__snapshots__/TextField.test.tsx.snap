// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render the TextField 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.8rem;
  width: 100%;
}

.c1 {
  color: #696F8C;
  font-size: 1.2rem;
  line-height: 1.6rem;
}

.c1::before {
  color: #D14343;
}

.c2 {
  background-color: #FFFFFF;
  border-radius: 0.4rem;
  font-size: 1.2rem;
  padding: 1.2rem 0.8rem;
  width: 100%;
  border: 1px solid;
  border-color: #D8DAE5;
  color: #8F95B2;
}

.c2:focus {
  border-color: #9DB5FF;
  box-shadow: 0 0 0 2px #D6E0FF;
  outline-color: #9DB5FF;
}

.c2:disabled {
  background-color: #F4F6FA;
  border: 1px solid #E6E8F0;
  color: #8F95B2;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="description"
  >
    Descrição
  </label>
  <input
    class="c2"
    id="description"
    name="description"
    type="text"
  />
</div>
`;
